python/pfs/drp/stella/psf.py

```diff
@@ -0,0 +1,251 @@
+from abc import ABC, abstractmethod
+
+from lsst.afw.image import positionToIndex
+from lsst.afw.math import offsetImage
+from lsst.afw.fits import MemFileManager
+from lsst.afw.detection import Psf
+
+__all__ = ["SpectralPsf", "ImagePsf"]
+
+
+class SpectralPsf(ABC):
+    """Base class for PSF on a spectral image
+
+    This is based on the lsst::afw::detection::Psf class; we could inherit from
+    that class, but reimplementation in python is easy, and it allows us to
+    side-step all the hassles of C++. Also, the main calculation methods
+    (``computeImage``, ``computeKernelImage`` and ``computeShape``) take
+    ``fiberId,wavelength`` arguments; there are corresponding methods
+    (``computeImageXY``, ``computeKernelImageXY`` and ``computeShapeXY``) that
+    work in terms of position on the image.
+
+    Parameters
+    ----------
+    detectorMap : `pfs.drp.stella.DetectorMap`
+        Mapping from fiberId,wavelength <--> x,y.
+    warpAlgorithm : `str`
+        Warp kernel to use in ``computeImageXY``.
+    warpBuffer : `int`
+        Buffer around image to use ``computeImageXY``.
+    """
+    def __init__(self, detectorMap, warpAlgorithm="lanczos5", warpBuffer=5):
+        self.detectorMap = detectorMap
+        self.warpAlgorithm = warpAlgorithm
+        self.warpBuffer = warpBuffer
+
+    def computeImage(self, fiberId, wavelength):
+        """Return an image in the same coordinate system as the pixelated image
+
+        This is appropriate for fitting or subtracting a PSF.
+
+        Parameters
+        ----------
+        fiberId : `int`
+            Fiber identifier.
+        wavelength : `float`
+            Wavelength.
+
+        Returns
+        -------
+        image : `lsst.afw.image.Image`
+            Image of the PSF.
+        """
+        return self.computeImageXY(self.detectorMap.findPoint(fiberId, wavelength))
+
+    def computeImageXY(self, point):
+        """Return an image in the same coordinate system as the pixelated image
+
+        This is appropriate for fitting or subtracting a PSF.
+
+        Parameters
+        ----------
+        point : `lsst.geom.Point2D`
+            Image position at which to calculate.
+
+        Returns
+        -------
+        image : `lsst.afw.image.Image`
+            Image of the PSF.
+        """
+
+        kernel = self.computeKernelImageXY(point)
+        xWhole, xFrac = positionToIndex(point.getX(), True)
+        yWhole, yFrac = positionToIndex(point.getY(), True)
+        if xFrac != 0.0 or yFrac != 0.0:
+            image = offsetImage(kernel, xFrac, yFrac, self.warpAlgorithm, self.warpBuffer)
+        image.setXY0(xWhole + image.getX0(), yWhole + image.getY0())
+        return image
+
+    def computeKernelImage(self, fiberId, wavelength):
+        """Return an image with the PSF centered at 0,0 (when xy0 is respected)
```

Comment:
What is meant by 'when xy0 is respected'?

Reply:
The `lsst.afw.image.Image` class (and relatives) use two coordinate systems:
* `LOCAL`: this is the coordinate system of the image in memory. The lower-left pixel is `0,0`.
* `PARENT`: this is a logical coordinate system that is offset from `LOCAL` by integer values `x0,y0` (often abbreviated to `xy0`).

So the above statement means that the PSF may not be centered at the lower-left corner (`0,0` in the `LOCAL` frame), but the location of the center of the PSF is `0,0` in the `PARENT` frame.
This is the same language used in [`lsst::afw::detection::Psf::computeKernelImage`](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1afw_1_1detection_1_1_psf.html#ab6d8cec1248cb14247bd9d897f13597b).